src/persist-client/examples/persistent_source.rs

```diff
@@ -0,0 +1,568 @@
+    use tracing::error;
+
+    use crate::api::{Batch, Source, TimestampedBatch, Timestamper};
+
+    /// Spawns tokio tasks for the different steps of a source pipeline. We read from the source,
+    /// decode, assign timestamps to updates, and emit them.
+    pub async fn spawn_source_pipeline<S, TS, T>(
+        name: String,
+        source: S,
+        timestamper: TS,
+        batch_size: usize,
+    ) -> Result<(), Box<dyn Error>>
+    where
+        S: Source<SourceTimestamp = T> + Send + 'static,
+        TS: Timestamper<SourceTimestamp = T> + Send + Sync + 'static,
+        T: timely::PartialOrder + PartialOrd + Copy + Debug + Send + Sync + 'static,
+        S::SourceTimestamp: Send,
+        S::D: Send,
+        TS::SourceTimestamp: Send,
+        TS::AssignedTimestamp: Send,
+    {
+        // Right now, everything is a separate task, including the prefetching. We'd have to see
+        // how to split these in a real implementation.
+
+        let (raw_batches_tx, raw_batches_rx) = mpsc::channel(2);
+
+        let source_task = tokio::task::spawn(async move {
+            read_source(source, batch_size, raw_batches_tx).await;
+        });
+
+        let (raw_batches_tx2, raw_batches_rx2) = mpsc::channel(2);
+
+        let timestamper1 = timestamper.clone();
+        let prefetcher_task = tokio::task::spawn(async move {
+            prefetch_timestamp_bindings(timestamper1, raw_batches_rx, raw_batches_tx2).await;
+        });
+
+        let (decoded_batches_tx, decoded_batches_rx) = mpsc::channel(2);
+
+        let decoder_task = tokio::task::spawn(async move {
+            decode(raw_batches_rx2, decoded_batches_tx).await;
+        });
+
+        let (timestamped_batches_tx, timestamped_batches_rx) = mpsc::channel(2);
+
+        let timestamper_task = tokio::task::spawn(async move {
+            assign_timestamp(timestamper, decoded_batches_rx, timestamped_batches_tx).await;
+        });
+
+        let consumer_task = tokio::task::spawn(async move {
+            consume(name, timestamped_batches_rx).await;
+        });
+
+        let _ = tokio::join!(
+            source_task,
+            prefetcher_task,
+            decoder_task,
+            timestamper_task,
+            consumer_task
+        );
+
+        Ok(())
+    }
+
+    /// Reads batches from the given `source` and forwards them to `tx`.
+    async fn read_source<S, T>(
+        mut source: S,
+        batch_size: usize,
+        tx: mpsc::Sender<Batch<S::D, S::SourceTimestamp>>,
+    ) where
+        S: Source<SourceTimestamp = T>,
+        T: timely::PartialOrder + PartialOrd + Debug,
+    {
+        loop {
+            let batch = source.read_batch(batch_size).await;
+
+            if let Some(batch) = batch {
+                if let Err(_) = tx.send(batch).await {
+                    error!("receiver dropped");
+                    return;
+                }
+            } else {
+                tokio::time::sleep(Duration::from_millis(10)).await;
```

Comment:
nit: why does read_batch return an Option? this loop is simpler if it always waits until it can return a batch

Reply:
I had a thing in my head where sources would say "I don't have data right now, but I might in the future", and the `Option` is communicating that. Though in an async/async task world that might not be necessary.